File: 912. Sort an Array.java
class Solution {
    public List<Integer> sortArray(int[] a) {
        return Arrays.stream(mergeSort(a)).boxed().collect(Collectors.toList());
    }
    
    public int[] mergeSort(int[] a) {
        if (a.length <= 1) {
            return a;
        }
        
        int mid = a.length / 2;
		// divide section
        int[] left = mergeSort(Arrays.copyOfRange(a, 0, mid));
        int[] right = mergeSort(Arrays.copyOfRange(a, mid, a.length));
		// merge section
        for (int i = 0, j = 0; i < left.length || j < right.length;) {
		    // if our right side is already expended, or this left[i] is smaller, then use left[i].
            if (j == right.length || i < left.length && left[i] < right[j]) {
                a[i + j] = left[i];
                i++;
            }
            else {
                a[i + j] = right[j];
                j++;
            }
        }
        
        return a;
    }
}

File: 912. Sort an Array.kt
class Solution {
    fun sortArray(nums: IntArray): IntArray {
        if (nums.size <= 1) return nums
        val mid = nums.size / 2
        val left = sortArray(nums.sliceArray(0 until mid))
        val right = sortArray(nums.sliceArray(mid until nums.size))
        return merge(left, right)
    }

    fun merge(left: IntArray, right: IntArray): IntArray {
        var i = 0
        var j = 0
        val merged = IntArray(left.size + right.size)
        for (k in merged.indices) {
            when {
                j >= right.size -> merged[k] = left[i++]
                i >= left.size -> merged[k] = right[j++]                
                left[i] <= right[j] -> merged[k] = left[i++]
                else -> merged[k] = right[j++]
            }
        }
        return merged
    }
}
